### PR TITLE
Adopt lead pressure calculation algorithm

### LINKED ISSUE #30049: Adopt lead pressure calc algorithm (PEARL)
Function to find the pressure given the d-spacing of the lead (111) peak using the equation of state in [1] (I think).

[1] Strässle, Th, et al. "Equation of state of lead from high-pressure neutron diffraction up to 8.9 GPa and its implication for the NaCl pressure scale." Physical Review B 90.1 (2014): 014101.

https://journals.aps.org/prb/pdf/10.1103/PhysRevB.90.014101
```
from mantid.kernel import *
from mantid.api import *
from mantid.simpleapi import *
import numpy as np

class LeadPressureCalc(PythonAlgorithm):
    
    def category(self):
        return 'PEARL tools'
    def PyExec(self):
        # Run the algorithm
        pass
    def PyInit(self):
        self.declareProperty('d(111)', 2.8589, FloatBoundedValidator(lower=2, upper=2.8590), doc="Position of (111) reflection in dSpacing")
        self.declareProperty('T', 300, doc="Sample temperature in K")    
        self.declareProperty(MatrixWorkspaceProperty("OutputWorkspace",defaultValue="",direction=Direction.Output))
    def PyExec(self):
        v_calc = (np.sqrt(3) * (self.getProperty("d(111)").value))**3
        T = self.getProperty("T").value
        v00 =121.41813
        a=1.0582E-02
        b=3.4934E-06
        k00=41.72535
        c=-2.5437E-02
        d=-2.7567E-06
        k00p=5.3944
        e=0.0011
        pow = 0.3333333333333333333
        P_calc = 1.5*(k00+(c*(T-300))+(d*(T-300)**2))*((((v00+(a*(T-300))+(b*(T-300)**2))/v_calc)**(7*pow))-(((v00+(a*(T-300))+(b*(T-300)**2))/v_calc)**(5*pow)))*(1+0.75*((k00p+(e*T))-4)*((((v00+(a*(T-300))+(b*(T-300)**2))/v_calc)**(2*pow))-1))
        self.log().information("The pressure is " +  str(P_calc) + " GPa")
        	
        CreateSingleValuedWorkspace(P_calc)
        #print(alg["P_calc"])
        self.setProperty("OutputWorkspace", CreateSingleValuedWorkspace(P_calc))
        
AlgorithmFactory.subscribe(LeadPressureCalc)
```

### PR DESCRIPTION
**Pull request by clayton-tom**
**Description of work.**
Onboard a new algorithm from Chris Ridley and add some unit tests. The purpose of this algorithm is to take the dSpacing of the 111 bragg peak and the temperature, and calculate the sample pressure using an equation of state. It can also, if supplied with a target pressure, try a range of dSpacing values to see which corresponds to the target pressure for the input temperature.
Original file: [LeadPressureCalc.zip](https://github.com/mantidproject/mantid/files/5608726/LeadPressureCalc.zip)

**Report to:** Chris Ridley

**To test:**
1. Run the algorithm with a range of different values of dSpacing and temperature in the valid ranges to ensure sensible answers are returned
2. Run the algorithm with input values out of range to ensure the input validation catches it

Fixes #30049 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by samtygier-stfc** [comment]
I think you could avoid the loop and do the whole thing in numpy. It would loose the optimisation of trimming down to candidates below TOL, and mean that the whole array of 10k values is held in memory. But I suspect it would be faster. Something like: 

```
test_dspacing = np.arange(2, 2.95, 0.0001)
pressure = calculate_pressure(test_dspacing, temp)
diff = abs(pressure - p_target)
i = np.argmin(diff)
diff, found_d = diff[i], test_dspacing[i]

if diff < TOL:
	if use_input_target:
	...
```

**Comment by clayton-tom** [comment]
> I think you could avoid the loop and do the whole thing in numpy. It would loose the optimisation of trimming down to candidates below TOL, and mean that the whole array of 10k values is held in memory. But I suspect it would be faster. Something like:
> 
> ```
> test_dspacing = np.arange(2, 2.95, 0.0001)
> pressure = calculate_pressure(test_dspacing, temp)
> diff = abs(pressure - p_target)
> i = np.argmin(diff)
> diff, found_d = diff[i], test_dspacing[i]
> 
> if diff < TOL:
> 	if use_input_target:
> 	...
> ```

@samtygier-stfc thanks for your comment, I've given it a go and it's certainly faster (~0.02 vs ~0.09 seconds to run exec for same inputs), and I think there probably aren't many mantid users who will struggle with the ~29000 ndarray values so will go ahead with this

File: Framework/PythonInterface/test/python/plugins/algorithms/LeadPressureCalcTest.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+#
+# Copyright &copy; 2020 ISIS Rutherford Appleton Laboratory UKRI,
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+import unittest
+from mantid.simpleapi import mtd, LeadPressureCalc
+
+
+class MyTestCase(unittest.TestCase):
```
Comment (samtygier-stfc) [inline Framework/PythonInterface/test/python/plugins/algorithms/LeadPressureCalcTest.py:11]: Should this be called `LeadPressureCalcTest`

**Review by samtygier-stfc — CHANGES_REQUESTED** [review]
Code looks good. Test case and example work. Probably someone else should check that the answers are sensible.

Would it be normal for the table workspace to be returned as an output? For example so I could do

`lead_calc = LeadPressureCalc(dSpacing=2.85, T=312, TargetPressure=0.0043)`

**Review by samtygier-stfc — APPROVED** [review]

**Comment by samtygier-stfc** [comment]
Looks good to me.

**Comment by DannyHindson** [comment]
@clayton-tom - well done on completing this Tom. Please can you check the nightly build succeeds tonight and assuming it does, can you email Chris and ask him to have a look?

**Pull request merged**